### ISSUE TITLE
idea: compare avatars

### ISSUE THREAD

**Issue by tobozo** [kawamataryo/sky-follower-bridge#13]
hi, 

The possibility of false positive matching with short account names (e.g. less than 5 characters) is superior to 10%. 

This number can eventually be lowered by comparing avatar images between twitter and bluesky, giving more confidence to the matching evaluation.

However, similar images may still differ (scaling, cropping), so this is a double-edged situation.

https://github.com/rsmbl/Resemble.js
https://github.com/mapbox/pixelmatch

**Comment by kawamataryo** [comment]
Thank you for the suggestion. It's a good idea.
There may be some issues such as differences in image size and performance of comparison processing, but I will consider if it is implementable.

**Comment by cooljeanius** [comment]
I was thinking of this, too, as it's one of the steps that I take manually

**Comment by Zero3K** [comment]
Any news regarding this?

**Comment by cooljeanius** [comment]
@PropertyOfMyCat had some ideas relevant to this in issue #33

**Comment by PropertyOfMyCat** [comment]
For avatar matching, I propose first preprocessing the images from both Twitter and Bluesky to account for differences that are artefacts that might result from differences in image compression. Probably the route to go could be:

1. Greyscale
2. Threshold to create B/W image
3. Reduce pixel resolution to something common to both (side-effect: should speed-up processing)
4. Compare with [pixelmatch](https://www.npmjs.com/package/pixelmatch) (simple) or [opencv](https://docs.opencv.org/4.x/d5/d10/tutorial_js_root.html) (more complicated, likely better results)
5. Check if similarity score passes threshold to infer a match/mismatch 

Alternatively, just show both avatars side-by-side in the results and let the human verify that the match isn't a false positive!

**Comment by tobozo** [comment]
palette analysis could also help with identifying default avatar images so that their score can be lowered

![image](https://github.com/user-attachments/assets/af4b62e4-6bc9-4dfc-92c9-ecea5eaeb85d)

**Comment by kawamataryo** [comment]
I am currently working on this feature. There is a possibility of false negatives, but it seems feasible to some extent. Please look forward to the release.

**Comment by tobozo** [comment]
@kawamataryo you might be interested by this project https://github.com/ISCPIF/OpenPortability

theyr'e running a site that should eventually help with migrating the followers by using account matching https://openportability.org

**Comment by kawamataryo** [comment]
@tobozo That's website was previously called HelloQuiteX, wasn't it? I signed up, but it didn't work properly. Did you manage to transfer your account?

**Comment by tobozo** [comment]
yup that's HelloQuitteX, they had to move to a hosting with better DDOS protection, and they should enable the migration tools any time soon.

**Comment by tobozo** [comment]
@kawamataryo the migration tool was finally enabled, it only found 29 more accounts but doesn't specify if those are followers, followees, or both, I suspect it is both so I can't tell if/why sky follower bridge missed those accounts.

![Image](https://github.com/user-attachments/assets/ba4ab5b3-6196-4ae9-93a6-3389a71e5085)

**Comment by kawamataryo** [comment]
@tobozo 
I see! I'll check it out myself.

I think OpenPortability matches using a different logic than Sky Follower Bridge. Since the matching logic of Sky Follower Bridge is not perfect, discrepancies are unavoidable.

**Comment by kawamataryo** [comment]
Avatar comparison was implemented in v2.8.0. Please give it a try.
https://bsky.app/profile/sky-follower-bridge.dev/post/3lheppj4jlc2s
<img width="786" alt="Image" src="https://github.com/user-attachments/assets/111ff80a-707e-420f-a701-520e34653fcf" />
